File: apps/api/app/schemas/distribuidor.py
"""
Schemas de distribuidor.
"""

from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from app.schemas.categoria import CategoriaResponse

# ── Perfil y Configuración ────────────────────────────────────────

class DistribuidorResponse(BaseModel):
    """Información pública o privada del distribuidor."""
    id: UUID
    nombre: str
    nombre_negocio: str
    rfc: str
    direcciones: list["DireccionDistribuidorResponse"]
    imagen_perfil: str | None
    imagen_fondo: str | None
    valoracion_promedio: float
    total_valoraciones: int
    fecha_creacion: datetime | None
    categorias: list[CategoriaResponse] | None = None

    model_config = {"from_attributes": True}


class DistribuidorUpdateInfo(BaseModel):
    """Campos permitidos para actualización por el distribuidor."""
    nombre_negocio: str | None = None
    telefono: str | None = None
    categorias: list[UUID] | None = None

# ── Direcciones ───────────────────────────────────────────────────

# RegistroDireccionDistribuidorRequest -> esta en schema/auth porque se me antojo algun problema?????!!!

class DireccionDistribuidorResponse(BaseModel):
    """Información de la dirección física del distribuidor."""
    id: UUID
    calle: str
    ciudad: str
    estado: str
    codigo_postal: str
    es_predeterminada: bool

    model_config = {"from_attributes": True}



class UpsertDireccionDistribuidorRequest(BaseModel):
    """Datos para crear o actualizar una dirección del distribuidor."""
    calle: str
    ciudad: str
    estado: str
    codigo_postal: str
    es_predeterminada: bool = False


# ── Catálogo de Productos del Distribuidor ────────────────────────

class ProductoCatalogoResponse(BaseModel):
    """Producto en el catálogo minificado."""
    producto_id: UUID
    nombre: str
    costo: float
    disponible: bool
    unidad: str
    existencias: int
    imagen: str | None = None


class CatalogoPaginatedResponse(BaseModel):
    """Respuesta paginada del catálogo de un distribuidor."""
    total_productos: int
    total_paginas: int
    pagina_actual: int
    tiene_siguiente: bool
    tiene_anterior: bool
    siguiente_url: str | None = None
    anterior_url: str | None = None
    productos: list[ProductoCatalogoResponse]


# ── Valoraciones ──────────────────────────────────────────────────

class ValoracionResponse(BaseModel):
    """Respuesta de una valoración de un distribuidor."""
    id: UUID
    cliente_id: UUID
    puntuacion: int
    comentario: str | None
    created_at: datetime




# ── Catálogo Global de Distribuidores ─────────────────────────────

class MiniDistribuidorResponse(BaseModel):
    """Respuesta minificada de un distribuidor para el catálogo."""
    distribuidor_id: UUID
    nombre_negocio: str
    imagen_fondo: str | None
    valoracion_promedio: float | None
    total_valoraciones: int | None
    categorias: list[str] | None


class DistribuidoresPaginatedResponse(BaseModel):
    """Respuesta paginada del catálogo de distribuidores."""
    total_distribuidores: int
    total_paginas: int
    pagina_actual: int
    tiene_siguiente: bool
    tiene_anterior: bool
    siguiente_url: str | None = None
    anterior_url: str | None = None
    distribuidores: list[MiniDistribuidorResponse]


# ── Dashboard y Alertas ───────────────────────────────────────────

class ResumenDashboardResponse(BaseModel):
    """Resumen para el dashboard del distribuidor."""
    volumen_bruto_mes: float
    pedidos_activos: int
    productos_poco_stock: int


class AlertaExistenciaResponse(BaseModel):
    """Alerta de existencias bajas para un producto."""
    producto_id: UUID
    nombre: str
    sku: str
    existencias: int
    costo: float
    unidad: str
    imagen: str | None = None
    disponible: bool
    estado_stock: str


class PedidoActivoDistribuidorResponse(BaseModel):
    """Información de un pedido activo de un distribuidor."""
    pedido_id: UUID
    orden_id: UUID
    cliente_nombre: str
    cliente_email: str
    total: float
    estado: str
    confirmado_at: datetime

File: packages/shared/src/api/distribuidor.ts
import { fetchWithAuth } from "./fetch";
import { API_URL } from "../constants";

// ─── Tipos públicos del mercado ──────────────────────────────────────────────

export interface MiniDistribuidorResponse {
    distribuidor_id: string;
    nombre_negocio: string;
    imagen_fondo: string | null;
    valoracion_promedio: number | null;
    total_valoraciones: number | null;
    categorias: string[] | null;
}

export interface DistribuidoresPaginatedResponse {
    total_distribuidores: number;
    total_paginas: number;
    pagina_actual: number;
    tiene_siguiente: boolean;
    tiene_anterior: boolean;
    siguiente_url: string | null;
    anterior_url: string | null;
    distribuidores: MiniDistribuidorResponse[];
}

export interface DistribuidorPublicoResponse {
    id: string;
    nombre: string;
    nombre_negocio: string;
    rfc: string;
    imagen_perfil: string | null;
    imagen_fondo: string | null;
    valoracion_promedio: number;
    total_valoraciones: number;
    fecha_creacion: string | null;
    categorias: { id: string; nombre: string; imagen: string | null }[] | null;
    direcciones: {
        id: string;
        calle: string;
        ciudad: string;
        estado: string;
        codigo_postal: string;
        es_predeterminada: boolean;
    }[];
    es_verificado: boolean;
    descripcion: string | null;
}

export interface ProductoCatalogoPublico {
    producto_id: string;
    nombre: string;
    costo: number;
    disponible: boolean;
    unidad: string;
    existencias: number;
    imagen: string | null;
}

export interface CatalogoDistribuidorPublico {
    total_productos: number;
    total_paginas: number;
    pagina_actual: number;
    tiene_siguiente: boolean;
    tiene_anterior: boolean;
    siguiente_url: string | null;
    anterior_url: string | null;
    productos: ProductoCatalogoPublico[];
}

export interface ResumenMensual {
    volumen_bruto_mes: number;
    pedidos_activos: number;
    productos_poco_stock: number;
}

export interface AlertaExistencia {
    producto_id: string;
    nombre: string;
    sku: string;
    existencias: number;
    costo: number;
    unidad: string;
    imagen: string | null;
    disponible: boolean;
    estado_stock: string;
}

export interface PedidoActivo {
    pedido_id: string;
    orden_id: string;
    cliente_nombre: string;
    cliente_email: string;
    total: number;
    estado: string;
    confirmado_at: string;
}

// ─── Funciones autenticadas ──────────────────────────────────────────────────

export const obtenerResumenMensual = async (
    token?: string,
    umbral_stock: number = 67
): Promise<ResumenMensual | null> => {
    const respuesta = await fetchWithAuth(`/distribuidores/me/resumen?umbral_stock=${umbral_stock}`, {
        method: 'GET',
    }, token);

    if (respuesta.ok) {
        return await respuesta.json();
    }
    return null;
}

/**
 * `archivo` es un `File`/`Blob` en web; en Expo se pasa el descriptor
 * `{ uri, name, type }` que acepta el `FormData` de React Native.
 */
export async function actualizarImagenNegocio(
    distribuidorId: string,
    archivo: Blob,
    token?: string
): Promise<boolean> {
    const formData = new FormData();
    formData.append('file', archivo);

    const respuesta = await fetchWithAuth(`/distribuidores/${distribuidorId}/imagen-negocio`, {
        method: 'POST',
        body: formData
    }, token);

    return respuesta.ok;
}

export const obtenerProductosPocasExistencias = async (
    token?: string,
    umbral_stock: number = 67
): Promise<AlertaExistencia[]> => {
    const respuesta = await fetchWithAuth(`/distribuidores/me/alertas-existencias?umbral_stock=${umbral_stock}`, {
        method: 'GET',
    }, token);

    if (respuesta.ok) {
        return await respuesta.json();
    }
    return [];
}

export const obtenerPedidosActivos = async (token?: string): Promise<PedidoActivo[]> => {
    const respuesta = await fetchWithAuth(`/distribuidores/me/pedidos-activos`, {
        method: 'GET',
    }, token);

    if (respuesta.ok) {
        return await respuesta.json();
    }
    return [];
}

// ─── Funciones públicas del mercado (sin auth) ────────────────────────────────

/**
 * Lista distribuidores de forma paginada. Público, sin autenticación requerida.
 */
export async function listarDistribuidores(
    pagina: number = 1,
    cantidad: number = 12,
    categorias?: string[],
    valoracionMin?: number,
    valoracionMax?: number,
): Promise<DistribuidoresPaginatedResponse> {
    const params = new URLSearchParams({
        numero_pagina: String(pagina),
        cantidad_pagina: String(cantidad),
    });
    if (valoracionMin !== undefined) params.set("valoracion_min", String(valoracionMin));
    if (valoracionMax !== undefined) params.set("valoracion_max", String(valoracionMax));
    if (categorias && categorias.length > 0) {
        categorias.forEach((c) => params.append("categorias", c));
    }

    try {
        const res = await fetch(`${API_URL}/distribuidores/?${params.toString()}`, {
            cache: "no-store",
        });
        if (res.ok) return await res.json();
    } catch (e) {
        console.error("Error listando distribuidores:", e);
    }
    return {
        total_distribuidores: 0,
        total_paginas: 0,
        pagina_actual: 1,
        tiene_siguiente: false,
        tiene_anterior: false,
        siguiente_url: null,
        anterior_url: null,
        distribuidores: [],
    };
}

/**
 * Obtiene el perfil completo de un distribuidor. Público, sin autenticación.
 */
export async function obtenerDistribuidor(id: string): Promise<DistribuidorPublicoResponse | null> {
    try {
        const res = await fetch(`${API_URL}/distribuidores/other/${id}`, { cache: "no-store" });
        if (res.ok) return await res.json();
    } catch (e) {
        console.error("Error obteniendo distribuidor:", e);
    }
    return null;
}

/**
 * Obtiene el catálogo de productos de un distribuidor. Público, sin autenticación.
 */
export async function obtenerCatalogoDistribuidorPublico(
    distribuidorId: string,
    pagina: number = 1,
    cantidad: number = 20,
): Promise<CatalogoDistribuidorPublico> {
    const params = new URLSearchParams({
        numero_pagina: String(pagina),
        cantidad_pagina: String(cantidad),
    });
    try {
        const res = await fetch(
            `${API_URL}/distribuidores/${distribuidorId}/catalogo?${params.toString()}`,
            { cache: "no-store" }
        );
        if (res.ok) return await res.json();
    } catch (e) {
        console.error("Error obteniendo catálogo del distribuidor:", e);
    }
    return {
        total_productos: 0,
        total_paginas: 0,
        pagina_actual: 1,
        tiene_siguiente: false,
        tiene_anterior: false,
        siguiente_url: null,
        anterior_url: null,
        productos: [],
    };
}
